x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java

```diff
@@ -222,14 +222,20 @@ private boolean checkIndexAvailable(ClusterState state) {
     }
 
     private boolean checkIndexMappingUpToDate(ClusterState clusterState) {
+        final Version minimumNonClientNodeVersion = clusterState.nodes().getSmallestNonClientNodeVersion();
+        final SystemIndexDescriptor descriptor = systemIndexDescriptor.getDescriptorCompatibleWith(minimumNonClientNodeVersion);
```

Comment:
I had concerns about whether this could be an issue in a mixed environment when the following sequence of events happens:
1. An old node drops out
2. The cluster decides to upgrade the mappings because remaining nodes are new enough
3. The old node reconnects. It is possible for an old node to join a newer cluster because our doc says "older nodes **cannot always** join a cluster with a newer master".

At this point, I worried whether the upgraded mappings would cause failures for the old node. But it seems the index with upgraded mappings will never be allocated to the old node. _So I think we are good here_. I am writing my thoughts down in case I am missing something.

Reply:
I also believe that we are good in that scenario. Thanks for sharing your thoughts! It gives another person an opportunity to also think through the scenario.